exporter/awsemfexporter/metric_translator.go

```diff
@@ -118,6 +136,253 @@ func (dps DoubleHistogramDataPointSlice) At(i int) DataPoint {
+	// Retrieve Datapoints
+	var dps DataPoints
+	switch metric.DataType() {
+		case pdata.MetricDataTypeIntGauge:
+			dps = IntDataPointSlice{metric.IntGauge().DataPoints()}
+		case pdata.MetricDataTypeDoubleGauge:
+			dps = DoubleDataPointSlice{metric.DoubleGauge().DataPoints()}
+		case pdata.MetricDataTypeIntSum:
+			dps = IntDataPointSlice{metric.IntSum().DataPoints()}
+		case pdata.MetricDataTypeDoubleSum:
+			dps = DoubleDataPointSlice{metric.DoubleSum().DataPoints()}
+		case pdata.MetricDataTypeDoubleHistogram:
+			dps = DoubleHistogramDataPointSlice{metric.DoubleHistogram().DataPoints()}
+		default:
+			config.logger.Warn("Unhandled metric data type.",
+				zap.String("DataType", metric.DataType().String()),
+				zap.String("Name", metric.Name()),
+				zap.String("Unit", metric.Unit()),)
+			return
+	}
+
+	if dps.Len() == 0 {
+		return
+	}
+
+	for m := 0; m < dps.Len(); m++ {
+		dp := dps.At(m)
+		if dp.IsNil() {
+			continue
+		}
+
+		labelsMap := dp.LabelsMap()
+		// `labels` contains label key/value pairs
+		labels := make(map[string]string, labelsMap.Len()+1)
+		labelsMap.ForEach(func(k, v string) {
+			labels[k] = v
+		})
+
+		// Add OTel instrumentation lib name as an additional label if it is defined
+		if instrumentationLibName != noInstrumentationLibraryName {
+			labels[OTellibDimensionKey] = instrumentationLibName
+		}
+
+		key := getGroupedMetricKey(namespace, timestamp, labels)
+
+		if _, ok := groupedMetricMap[key]; ok {
+			updateGroupedMetric(dp, timestamp, labels, metric, key, groupedMetricMap, config)
+		} else {
+			groupedMetric := buildGroupedMetric(dp, namespace, timestamp, labels, metric, config.DimensionRollupOption)
+			if groupedMetric != nil {
+				groupedMetricMap[key] = groupedMetric
+				groupedMetrics = append(groupedMetrics,groupedMetric)
+			}
+		}
+	}
+	return
+}
+
+// getGroupedMetricKey generates a key for a given GroupedMetric
+func getGroupedMetricKey(cwNamespace string, timestamp int64, labels map[string]string) (string) {
+	var sb strings.Builder
+	keySlice := make([]string, 0, len(labels) + 2)
+	fields := make(map[string]string)
+	fields[namespaceKey] = cwNamespace
+	fields[timestampKey] = strconv.FormatInt(timestamp, 10)
+	keySlice = append(keySlice, namespaceKey, timestampKey)
+
+	for k, v := range labels {
+		fields[k] = v
+		keySlice = append(keySlice, k)
+	}
+
+	sort.Strings(keySlice)
+	keySliceLen := len(keySlice)
+
+	for i, j := range keySlice {
+		keyValuePair := j + ":" + fields[j]
+		sb.WriteString(keyValuePair)
+		if i < keySliceLen-1 {
+			sb.WriteString(",")
+		}
+	}
+	return sb.String()
+}
+
+// buildGroupedMetric builds GroupedMetric from Datapoint and pdata.Metric
+func buildGroupedMetric (dp DataPoint, namespace string, timestamp int64, labels map[string]string, pMetricData *pdata.Metric, dimensionRollupOption string) (*GroupedMetric) {
+	metrics := make(map[string]*MetricInfo)
+
+	// Extract metric
+	var metricVal interface{}
+	switch metric := dp.(type) {
+	case pdata.IntDataPoint:
```

Comment:
Instead of duplicating the `metric value calculation` logic in both buildGroupedMetric and updateGroupedMetric, please consolidate it in one place.

Reply:
Resolved in commit [8fa027b](https://github.com/hdj630/opentelemetry-collector-contrib/pull/95/commits/8fa027b16a00f14567a66aea4b9230381ffc946d)

Reply:
If you move it before calculating groupKey as @hdj630 said, you can combine both `buildGroupedMetric` and `updateGroupedMetric` into 1 function.